include/xtensor/utils/xutils.hpp

```diff
@@ -283,29 +283,31 @@ namespace xt
      * apply implementation *
      ************************/
 
-    namespace detail
-    {
-        template <class R, class F, std::size_t I, class... S>
-        R apply_one(F&& func, const std::tuple<S...>& s) NOEXCEPT(noexcept(func(std::get<I>(s))))
-        {
-            return static_cast<R>(func(std::get<I>(s)));
-        }
-
-        template <class R, class F, std::size_t... I, class... S>
-        R apply(std::size_t index, F&& func, std::index_sequence<I...> /*seq*/, const std::tuple<S...>& s)
-            NOEXCEPT(noexcept(func(std::get<0>(s))))
-        {
-            using FT = std::add_pointer_t<R(F&&, const std::tuple<S...>&)>;
-            static const std::array<FT, sizeof...(I)> ar = {{&apply_one<R, F, I, S...>...}};
-            return ar[index](std::forward<F>(func), s);
-        }
-    }
-
     template <class R, class F, class... S>
     inline R apply(std::size_t index, F&& func, const std::tuple<S...>& s)
-        NOEXCEPT(noexcept(func(std::get<0>(s))))
     {
-        return detail::apply<R>(index, std::forward<F>(func), std::make_index_sequence<sizeof...(S)>(), s);
+        return std::apply(
+            [&](const S&... args) -> R
+            {
+                return [&](const auto&... t) -> R
+                {
+                    auto f_impl = [&](auto&& self, auto&& i, auto&& h, auto&&... t) -> R
+                    {
+                        if (i == index)
+                        {
+                            return func(h);
+                        }
+                        if constexpr (sizeof...(t) > 0)
+                        {
+                            return self(self, std::size_t{i + 1}, t...);
+                        }
+                        XTENSOR_THROW(std::runtime_error, "Index applied to tuple of insufficient length");
```

Comment:
@JohanMabille I think the noexcept specification here was too restrictive. We need to provide some way of indicating when index exceeds sizeof...(S). In the old implementation I think we would simply get an access violation which is not ideal.

Reply:
The initial idea was to disable all checks to get the most opitmized generated code; I don't remember whether this function is used in critical parts of the code, so I cannot say for the NOEXCEPT, but I think we should prefer XTENSOR_ASSERT to XTENSOR_THROW.

Reply:
Moved to assert. Clang compiles this code nicely... MSVC was a bit of a mess but I'm not sure if it's a code generation issue or just noise.

Reply:
This is a link to the generated code. https://godbolt.org/z/sMdT517G4

It can calulate the result at compile time. Can I say it is the optimal solution then?

Reply:
Haha yeah I think so ;)